src/shell/shell_misc.cpp

```diff
@@ -389,106 +390,38 @@ void DOS_Shell::InputCommand(char * line) {
-	/* if not, then return without modifying "line" */
-	while (*r != 0 && *r != '%') r++;
-	if (*r != '%') return;
-	/* if the incoming string is already too long, then that's a problem too! */
-	if (((size_t)(r+1-line)) >= CMD_MAXLINE) {
-		LOG_MSG("DOS_Shell::ProcessCmdLineEnvVarStitution WARNING incoming string to substitute is already too long!\n");
-		goto overflow;
-	}
-	/* copy the string down up to that point */
-	for (const char* c = line; c < r;) {
-		assert(w < wf);
-		*w++ = *c++;
-	}
-	/* begin substitution process */
-	while (*r != 0) {
-		if (*r == '%') {
-			r++;
-			if (*r == '%' || *r == 0) {
-				/* %% or leaving a trailing % at the end (Win95 behavior) becomes a single '%' */
-				if (w >= wf) goto overflow;
-				*w++ = '%';
-				if (*r != 0) r++;
-				else break;
-			}
-			else {
-				const char* name = r; /* store pointer, 'r' is first char of the name following '%' */
-				int spaces = 0,chars = 0;
-				if (isalpha(*r) || *r == ' ') { /* must start with a letter. space is apparently valid too. (Win95) */
-					if (*r == ' ') spaces++;
-					else if (isalpha(*r)) chars++;
-					r++;
-					while (*r != 0 && *r != '%') {
-						if (*r == ' ') spaces++;
-						else chars++;
-						r++;
-					}
-				}
-				if (*r == '%' && ((spaces > 0 && chars == 0) || (spaces == 0 && chars > 0))) {
-					std::string temp2;
-					/* valid name found. substitute */
-					*r++ = 0; /* ASCIIZ snip */
-					if (GetEnvStr(name,temp2)) {
-						size_t equ_pos = temp2.find_first_of('=');
-						if (equ_pos != std::string::npos) {
-							const char *base = temp2.c_str();
-							const char *value = base + equ_pos + 1;
-							const char *fence = base + temp2.length();
-							assert(value >= base && value <= fence);
-							size_t len = (size_t)(fence-value);
-							if ((w+len) > wf) goto overflow;
-							memcpy(w,value,len);
-							w += len;
-						}
-					}
-				}
-				else {
-					/* nope. didn't find a valid name */
-					while (*r == ' ') r++; /* skip spaces */
-					name--; /* step "name" back to cover the first '%' we found */
-					for (const char* c = name; c < r;) {
-						if (w >= wf) goto overflow;
-						*w++ = *c++;
-					}
-				}
-			}
+/* Note: Buffer pointed to by "line" must be at least CMD_MAXLINE+1 bytes long! */
+void DOS_Shell::ProcessCmdLineEnvVarStitution(char* line) {
+	constexpr char surrogate_percent = 8;
+	const static std::regex re("\\%([^%0-9][^%]*)?%");
+	std::string text = line;
+	std::smatch match;
+	/* Iterates over potential %var1%, %var2%, etc matches found in the text string */
+	while (std::regex_search(text, match, re)) {
+		std::string variable_name;
+		variable_name = match[1].str();
+		if (!variable_name.size()) {
+			/* Replace %% with the character "surrogate_percent", then (eventually) % */
+			text.replace(match[0].first, match[0].second, std::string(1, surrogate_percent));
+			continue;
+		}
+		while (variable_name.size() && isspace(variable_name.front()))
+			variable_name.erase(0, 1);
```

Comment:
Sorry for this last change @Wengier ; I think we can range-drop to left-trim, and avoid the loop:

``` c++
variable_name.erase(0, variable_name.find_first_not_of(' '));
```

I added a quick test here, just to double check it handles a couple scenarios:
 - https://onlinegdb.com/ir8zxDDnz

This should be it!

Reply:
Thanks for the suggestion. I have modified the code accordingly.